### ISSUE TITLE
Dynamically generate policy-related content

### ISSUE THREAD

**Issue by kleinfreund** [kumahq/kuma-gui#341]
**TL;DR**: In order to allow Kuma to introduce and change existing policies more easily, Kuma GUI should dynamically generate policy-related content based on available policy types that are discovered via the `/policies` API endpoint.

## Problem statement

Currently, all policies in Kuma GUI are represented via independent and static content or components. Each policy overview page is its own component and these components are pretty much identical to each other safe for some minor differences (e.g. titles, labels, URLs to documentation, etc.).

This leads to a maintenance problem. Changing the UI for the policy overview pages requires changes to all 14 components (one for each policy type). Also, introducing a new policy in Kuma requires introducing one more such component in Kuma GUI and updating all related places in Kuma GUI that deal with showing general policy content (e.g. data plane proxy overviews).

## Solution

By leveraging the `/policies` API endpoint, it’s possible to make this completely dynamic.

### Dynamically generating Vue router routes (non-blocking)

It’s not possible to programmatically add new child routes to an existing route in vue-router@^3. One can only add new top level routes. For this reason, it’s necessary to _delay_ the instantiation of Vue and Vue router until the available policy types have been fetched from the `/policies` endpoint. This is a very quick operation, but in principle, one would like avoid delaying the app instantiation process if possible.

### Missing data on `/policies` endpoint (blocking)

The current state of the `/policies` endpoint is insufficient to completely remove all non-dynamic policy data from Kuma GUI. Specifically, the following things are missing or need to be changed:

- **Missing property `displayNameSingular`**: In some places of the UI, the singular form of the display name is necessary.
  → **Triage outcome**: A new property will be added to the `/policies` endpoint.
- **Missing property `routePath`**: The route paths in Kuma GUI use slightly different strings than what’s provided with the path property (e.g. meshgatewayroutes is the API path and mesh-gateway-routes is the route path). In order to protect people who have bookmarked policies from broken URLs, we should add a routePath property to the endpoint. Alternatively, the route paths in Kuma GUI can be changed. Also, to disambiguate both properties, I recommend renaming `path` to `apiPath`.
  → **Triage outcome**: The `path` property will be used to construct the route paths. We consider breaking these URLs acceptable compared to maintaining Kuma GUI information in Kuma which is awkward.
- **Missing property `docsPath`**: The policy views all have a link to the relevant Kuma documentation. In order to construct the docs URLs correctly, another property is needed because most documentation path segments for policies are singular (e.g. `health-check` instead of `health-checks`).
  → **Triage outcome**: The `path` property will be used to construct the docs URLs and redirects will be introduced on the docs website to redirect these URLs to their canonical version. It should be noted that the docs URLs use dash-delimited path segments with singular forms (e.g. `mesh-gateway-route`) instead of concatenated path segments and plural forms (e.g. `meshgatewayroutes`).
- **Extra property**: Some additional policy types (e.g. External Services, Virtual Outbounds) are present on the endpoint while not currently present in Kuma GUI. To not mix introducing new policies with this refactoring task, I recommend to remove these extra policy types from the response. (note that it’s entirely possible that I misunderstand why they should be there; I’m new to this).
  → **Triage outcome**: Virtual Outbounds was missing in Kuma GUI so it is being kept. External Services will be removed from the response.
- **Missing property `isExperimental`**: Some policy types are considered experimental (they display a warning message on the policy overview page). In order to know when to show this warning, this property should be added. Currently, it would need to be `true` for Mesh Gateways and Mesh Gateway Routes.
  → **Triage outcome**: A new property will be added to the `/policies` endpoint.

**Comment by jakubdyszkiewicz** [comment]
> Triage:
> displayNameSingular: we need to add it
> routePath: we will break it
> docsPath: we need to add redirects
> Extra property: "Virtual Outbounds" this is a bug that we don't have this in the GUI! Remove "External Services" from the response
> Let's add isExperimental

**Comment by kleinfreund** [comment]
In order to not make a new `docsPath` property necessary, we need the following redirects on the docs website kuma.io so we can generate docs URLs in the form `https://kuma.io/docs/{kumaDocsVersion}/policies/{apiPath}/` where `apiPath` matches the `path` property from the `/policies` endpoint response.

The following redirects are necessary:

- **From**: `https://kuma.io/docs/{kumaDocsVersion}/policies/circuit-breakers`
  **To**: `https://kuma.io/docs/{kumaDocsVersion}/policies/circuit-breaker`
- **From**: `https://kuma.io/docs/{kumaDocsVersion}/policies/fault-injections`
  **To**: `https://kuma.io/docs/{kumaDocsVersion}/policies/fault-injection`
- **From**: `https://kuma.io/docs/{kumaDocsVersion}/policies/health-checks`
  **To**: `https://kuma.io/docs/{kumaDocsVersion}/policies/health-check`
- **From**: `https://kuma.io/docs/{kumaDocsVersion}/policies/meshgateways`
  **To**: `https://kuma.io/docs/{kumaDocsVersion}/policies/meshgateway`
- **From**: `https://kuma.io/docs/{kumaDocsVersion}/policies/meshgatewayroutes`
  **To**: `https://kuma.io/docs/{kumaDocsVersion}/policies/meshgatewayroute`
- **From**: `https://kuma.io/docs/{kumaDocsVersion}/policies/proxytemplates`
  **To**: `https://kuma.io/docs/{kumaDocsVersion}/policies/proxy-template`
- **From**: `https://kuma.io/docs/{kumaDocsVersion}/policies/rate-limits`
  **To**: `https://kuma.io/docs/{kumaDocsVersion}/policies/rate-limit`
- **From**: `https://kuma.io/docs/{kumaDocsVersion}/policies/retries`
  **To**: `https://kuma.io/docs/{kumaDocsVersion}/policies/retry`
- **From**: `https://kuma.io/docs/{kumaDocsVersion}/policies/timeouts`
  **To**: `https://kuma.io/docs/{kumaDocsVersion}/policies/timeout`
- **From**: `https://kuma.io/docs/{kumaDocsVersion}/policies/traffic-logs`
  **To**: `https://kuma.io/docs/{kumaDocsVersion}/policies/traffic-log`
- **From**: `https://kuma.io/docs/{kumaDocsVersion}/policies/traffic-routes`
  **To**: `https://kuma.io/docs/{kumaDocsVersion}/policies/traffic-route`
- **From**: `https://kuma.io/docs/{kumaDocsVersion}/policies/traffic-traces`
  **To**: `https://kuma.io/docs/{kumaDocsVersion}/policies/traffic-trace`
- **From**: `https://kuma.io/docs/{kumaDocsVersion}/policies/virtual-outbounds`
  **To**: `https://kuma.io/docs/{kumaDocsVersion}/policies/virtual-outbound`

**Note**: These are the policies currently shown in Kuma GUI; however, the docs website documents more entries under the policies section. Perhaps it makes sense to generally add plural-to-singular redirects for things currently shown under the policies section.

**Comment by kleinfreund** [comment]
Supersedes https://github.com/kumahq/kuma-gui/issues/336.

**Comment by kleinfreund** [comment]
This issue is being implemented in https://github.com/kumahq/kuma-gui/pull/346. Its blockers are gone as the `/policies` endpoint as been updated in Kuma (https://github.com/kumahq/kuma/pull/4791). The only blocker remaining is https://github.com/kumahq/kuma-gui/issues/341#issuecomment-1209014843.

**Comment by kleinfreund** [comment]
This was implemented in https://github.com/kumahq/kuma-gui/pull/346.
